core/src/Cuda/Kokkos_Cuda_Parallel_MDRange.hpp

```diff
@@ -108,13 +111,18 @@ class ParallelFor<FunctorType, Kokkos::MDRangePolicy<Traits...>, Kokkos::Cuda> {
               (m_rp.m_upper[1] - m_rp.m_lower[1] + block.y - 1) / block.y,
               maxblocks[1]),
           1);
+      KOKKOS_ASSERT(grid.x > 0 && grid.x <= maxblocks[0]);
+      KOKKOS_ASSERT(grid.y > 0 && grid.y <= maxblocks[1]);
       CudaParallelLaunch<ParallelFor, LaunchBounds>(
           *this, grid, block, 0, m_rp.space().impl_internal_space_instance());
     } else if (RP::rank == 3) {
-      const dim3 block(m_rp.m_tile[0], m_rp.m_tile[1], m_rp.m_tile[2]);
-      KOKKOS_ASSERT(block.x > 0);
-      KOKKOS_ASSERT(block.y > 0);
-      KOKKOS_ASSERT(block.z > 0);
+      const dim3 block(
+          m_rp.m_tile[0], m_rp.m_tile[1],
+          ((m_rp.m_tile[2] > maxthreads[2]) ? maxthreads[2] : m_rp.m_tile[2]));
```

Comment:
The point is that we already accounted for the maximum limits for the first two dimensions earlier but not for the last one, right?
Maybe, just refactor into a local lambda as well?

Reply:
OK, will do, and I know how to do the other one, too...